File: packages/codex-thread-bridge/src/codex_thread_bridge/roles.py
"""Which pair a ROLE runs on, and why this file declares none of them itself.

The execution policy already asks two questions of every creation and every resume: was a model
and an effort stated, and does this host allow that pair. Neither is the question that was
actually being got wrong. CRW runs three levels -- a supervisor bound to an initiative, a parent
to a project, a child to an issue -- and each level is meant to run on a different pair. A project
parent created on someone else's model answered both existing questions correctly and was still
wrong, because nothing related the task's ROLE to the pair it was started on.

So this module adds the third question, IS IT THIS ROLE'S PAIR, and deliberately supplies no
answer to it. The role ids live here because two packages have to agree on them. The values do
not, for the same reason the allowlist does not: a pair written into code would be a second
source of truth competing with the operator's file, and it would answer the approval question on
a host that configured nothing -- which is exactly what PRESENCE_ONLY refuses to do. A role this
host's policy does not declare is refused, never defaulted.

The supervisor is the one role with no pair, and that is a rule rather than an omission. Its model
is the user's own selection, so the policy expresses it as {"expectation": "record"} and the
supervisor's authorization is whatever its own recorded settings say. An entry that tries to pin a
supervisor's model does not load at all. That is what makes "the parent policy is never propagated
to the supervisor" a mechanism instead of a sentence somebody has to remember, and it is worth the
asymmetry: every other role's pair is derived from policy, and a derived pair can be restored
safely where a chosen one cannot.
"""

# The three levels, spelled exactly as the relay's own scope bindings spell them. A conformance
# test in the relay asserts the two sets are equal, because the packages cannot import each
# other's vocabulary in both directions and a silent divergence here would mean a role that is
# enforced on one side and unknown on the other.
SUPERVISOR = "supervisor"
PARENT = "parent"
CHILD = "child"
ROLES = (SUPERVISOR, PARENT, CHILD)

ROLE_UNKNOWN = "execution_role_unknown"
ROLE_MISMATCH = "execution_role_mismatch"

# The supervisor's expectation. Not a pair, and not a wildcard either: it means the authorization
# is the one recorded for that task, from a source attributable to the user.
RECORD = "record"
PAIR = "pair"

ROLE_MAXIMUM = 64

_ENTRY_KEYS = {"model", "reasoningEffort", "expectation"}


class RoleExpectation:
    """What one role is expected to run on, and where that expectation comes from."""

    __slots__ = ("role", "model", "reasoning_effort", "expectation")

    def __init__(self, role, model=None, reasoning_effort=None, expectation=PAIR):
        self.role = role
        self.model = model
        self.reasoning_effort = reasoning_effort
        self.expectation = expectation

    def __eq__(self, other):
        return isinstance(other, RoleExpectation) and self.receipt() == other.receipt()

    def __repr__(self):
        return f"RoleExpectation({self.receipt()!r})"

    def receipt(self) -> dict:
        """What a caller is told about this role. Model ids are not secrets; paths are."""
        return {
            "role": self.role,
            "expectation": self.expectation,
            "model": self.model,
            "reasoningEffort": self.reasoning_effort,
        }


def _text(value, where, error):
    if not isinstance(value, str) or not value.strip() or len(value) > ROLE_MAXIMUM * 8:
        raise error(f"{where} must be a non-empty string")
    return value


def parse(declared, error) -> dict:
    """Read the policy file's roles section. Absent means no role is enforced on this host.

    The caller supplies the exception class it wants raised, which keeps this module from
    importing the policy module that imports it. Every refusal below stops the server at startup,
    exactly like every other unusable policy: a roles section that was configured and then
    silently ignored is the one outcome an operator would never detect.
    """
    if declared is None:
        return {}
    if not isinstance(declared, dict):
        raise error("roles must be a JSON object keyed by role id")
    parsed = {}
    for role, entry in declared.items():
        if role not in ROLES:
            raise error(f"{role!r} is not a role; supported are {sorted(ROLES)}")
        if not isinstance(entry, dict):
            raise error(f"role {role!r} must be an object")
        unknown = sorted(set(entry) - _ENTRY_KEYS)
        if unknown:
            raise error(
                f"role {role!r} has unknown keys {unknown}; supported are {sorted(_ENTRY_KEYS)}"
            )
        expectation = entry.get("expectation", PAIR)
        if expectation not in (PAIR, RECORD):
            raise error(f"role {role!r} expectation must be {PAIR!r} or {RECORD!r}")
        if role == SUPERVISOR:
            # Refused rather than ignored. A file that pins the supervisor's model was written by
            # someone who believes it will take effect, and loading it while discarding that
            # belief is worse than refusing to start.
            named = sorted({"model", "reasoningEffort"} & set(entry))
            if named:
                raise error(
                    f"role {SUPERVISOR!r} cannot declare {named}: its model and effort are the "
                    "user's own selection, so its expectation is the recorded authorization"
                )
            parsed[role] = RoleExpectation(role, expectation=RECORD)
            continue
        if expectation == RECORD:
            parsed[role] = RoleExpectation(role, expectation=RECORD)
            continue
        absent = sorted({"model", "reasoningEffort"} - set(entry))
        if absent:
            raise error(f"role {role!r} is missing {absent}")
        parsed[role] = RoleExpectation(
            role,
            _text(entry["model"], f"the model of role {role!r}", error),
            _text(entry["reasoningEffort"], f"the effort of role {role!r}", error),
        )
    return parsed

File: internal/bridge/execution/roles.go
package execution

import "fmt"

// Role names and expectations from roles.py.
const (
	Supervisor = "supervisor"
	Parent     = "parent"
	Child      = "child"
	Pair       = "pair"
	Record     = "record"
)

// roleNames is roles.ROLES in declaration order; checks that iterate roles follow it.
var roleNames = [...]string{Supervisor, Parent, Child}

const supportedRoles = "['child', 'parent', 'supervisor']"

// Role is roles.RoleExpectation. Model and Effort are empty for a supervisor.
type Role struct{ Model, Effort, Expectation string }

func (r Role) receipt(name string) map[string]any {
	return map[string]any{"role": name, "expectation": r.Expectation, "model": nullable(r.Model), "reasoningEffort": nullable(r.Effort)}
}

func isRole(value any) bool {
	for _, name := range roleNames {
		if value == any(name) {
			return true
		}
	}
	return false
}

// parseRoles mirrors roles.parse: absent means no role is enforced on this host.
func parseRoles(declared any) (map[string]Role, error) {
	parsed := map[string]Role{}
	if declared == nil {
		return parsed, nil
	}
	section, ok := declared.(*object)
	if !ok {
		return nil, &PolicyError{"roles must be a JSON object keyed by role id"}
	}
	for _, name := range section.keys {
		if !isRole(name) {
			return nil, &PolicyError{fmt.Sprintf("%s is not a role; supported are %s", repr(name), supportedRoles)}
		}
		entry, ok := section.values[name].(*object)
		if !ok {
			return nil, &PolicyError{fmt.Sprintf("role %s must be an object", repr(name))}
		}
		if err := only(entry, []string{"expectation", "model", "reasoningEffort"}, "role "+repr(name)); err != nil {
			return nil, err
		}
		role, err := parseRole(name, entry)
		if err != nil {
			return nil, err
		}
		parsed[name] = role
	}
	return parsed, nil
}

func parseRole(name string, entry *object) (Role, error) {
	var expectation any = Pair
	if name == Supervisor {
		expectation = Record
	}
	if entry.has("expectation") {
		expectation = entry.values["expectation"]
	}
	if expectation != any(Pair) && expectation != any(Record) {
		return Role{}, &PolicyError{fmt.Sprintf("role %s expectation must be 'pair' or 'record'", repr(name))}
	}
	if name == Supervisor && expectation != any(Record) {
		return Role{}, &PolicyError{"role 'supervisor' expectation must be 'record': its model and effort are the user's own selection, not something the policy pins"}
	}
	if name != Supervisor && expectation != any(Pair) {
		return Role{}, &PolicyError{fmt.Sprintf("role %s expectation must be 'pair': only 'supervisor' defers to the recorded authorization, and letting another role do so would exempt it from the role check", repr(name))}
	}
	if name == Supervisor {
		if named := entry.present("model", "reasoningEffort"); len(named) > 0 {
			return Role{}, &PolicyError{fmt.Sprintf("role 'supervisor' cannot declare %s: its model and effort are the user's own selection, so its expectation is the recorded authorization", repr(named))}
		}
		return Role{Expectation: Record}, nil
	}
	if absent := entry.absent("model", "reasoningEffort"); len(absent) > 0 {
		return Role{}, &PolicyError{fmt.Sprintf("role %s is missing %s", repr(name), repr(absent))}
	}
	model, err := identifier(entry.values["model"], "the model of role "+repr(name), Maximum)
	if err != nil {
		return Role{}, err
	}
	effort, err := identifier(entry.values["reasoningEffort"], "the effort of role "+repr(name), Maximum)
	if err != nil {
		return Role{}, err
	}
	return Role{Model: model, Effort: effort, Expectation: Pair}, nil
}
